primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,794 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trie Cache
+//!
+//! Provides an implementation of the [`TrieCache`](trie_db::TrieCache) trait.
+//! The implementation is split into three types [`SharedTrieCache`], [`LocalTrieCache`] and
+//! [`TrieCache`]. The [`SharedTrieCache`] is the instance that should be kept around for the entire
+//! lifetime of the node. It will store all cached trie nodes and values on a global level. Then
+//! there is the [`LocalTrieCache`] that should be kept around per state instance requested from the
+//! backend. As there are very likely multiple accesses to the state per instance, this
+//! [`LocalTrieCache`] is used to cache the nodes and the values before they are merged back to the
+//! shared instance. Last but not least there is the [`TrieCache`] that is being used per access to
+//! the state. It will use the [`SharedTrieCache`] and the [`LocalTrieCache`] to fulfill cache
+//! requests. If both of them don't provide the requested data it will be inserted into the
+//! [`LocalTrieCache`] and then later into the [`SharedTrieCache`].
+//!
+//! The [`SharedTrieCache`] is bound to some maximum number of bytes. It is ensured that it never
+//! runs above this limit. However as long as data is cached inside a [`LocalTrieCache`] it isn't
+//! taken into account when limiting the [`SharedTrieCache`]. This means that for the lifetime of a
+//! [`LocalTrieCache`] the actual memory usage could be above the allowed maximum.
+
+use crate::{Error, NodeCodec};
+use hash_db::Hasher;
+use lru::LruCache;
+use nohash_hasher::{BuildNoHashHasher, IntMap, IntSet};
+use parking_lot::{Mutex, MutexGuard, RwLock, RwLockReadGuard};
+use std::{
+	collections::{
+		hash_map::{DefaultHasher, Entry},
+		HashMap, HashSet,
+	},
+	hash::Hasher as _,
+	mem,
+	sync::Arc,
+};
+use trie_db::{node::NodeOwned, CachedValue};
+
+const LOG_TARGET: &str = "trie-cache";
+
+/// No hashing [`lru::LruCache`].
+///
+/// The key is an `u64` that is expected to be unique.
+type NoHashingLruCache<T> = lru::LruCache<u64, T, BuildNoHashHasher<u64>>;
+
+/// Get the key for the value cache.
+///
+/// The key is build by hashing the actual storage `key` and the `storage_root`.
+fn value_cache_get_key(key: &[u8], storage_root: &impl AsRef<[u8]>) -> u64 {
+	let mut hasher = DefaultHasher::default();
```

Comment:
should we use a crypto hash function and take the 8 first byte instead? I am not even sure of the collision resistance of the 8 over 32 bytes, actually if we manage collision explicitely it won't be needed.